app/services/hashtag_autocomplete_service.rb

```diff
@@ -89,84 +60,133 @@ class HashtagItem
     # and must be unique so it can be used for lookups via the #lookup
     # method above.
     attr_accessor :ref
+
+    # The relative URL for the resource that is represented by the autocomplete
+    # item, used for the cooked hashtags, e.g. /c/2/staff
+    attr_accessor :relative_url
+
+    def to_h
+      {
+        relative_url: self.relative_url,
+        text: self.text,
+        icon: self.icon,
+        type: self.type,
+        ref: self.ref,
+        slug: self.slug
+      }
+    end
   end
 
   def initialize(guardian)
     @guardian = guardian
   end
 
-  def lookup(slugs)
+  ##
+  # Finds resources of the provided types by their exact slugs, unlike
+  # search which can search partial names, slugs, etc. Used for cooking
+  # fully formed #hashtags in the markdown pipeline. The @guardian handles
+  # permissions around which results should be returned here.
+  #
+  # @param {Array} slugs The fully formed slugs to look up, which can have
+  #                      ::type suffixes attached as well (e.g. ::category),
+  #                      and in the case of categories can have parent:child
+  #                      relationships.
+  # @param {Array} types_in_priority_order The resource types we are looking up
+  #                                        and the priority order in which we should
+  #                                        match them if they do not have type suffixes.
+  # @returns {Hash} A hash with the types as keys and an array of HashtagItem that
+  #                 matches the provided slugs.
+  def lookup(slugs, types_in_priority_order)
     raise Discourse::InvalidParameters.new(:slugs) if !slugs.is_a?(Array)
+    raise Discourse::InvalidParameters.new(:order) if !types_in_priority_order.is_a?(Array)
 
-    all_slugs = []
-    tag_slugs = []
+    types_in_priority_order =
+      types_in_priority_order.select { |type| @@data_sources.keys.include?(type) }
```

Comment:
something to be careful here, we don't any check on key being a string or a symbol from the beginning of the chain. I didn't check if that could cause issues, but we might want to be stricter here

Reply:
Okay will add more checks

Reply:
one solution could just be to convert at boundaries to either string or simbols